pep/003-dependency-namespaces-and-labels.md

```diff
@@ -0,0 +1,508 @@
+# List bundles in the index
+
+$ porter bundles list --name mysql
+# Bundles that have mysql in the name
+
+$ porter bundles list --label app=mysql
+# Bundles that have the label app=mysql
+
+$ porter bundles list --interface io.cnab.interface.mysql
+# Bundles that implement io.cnab.interface.mysql
+# Display a * on the line of a bundle that is the default implementation
+
+$ porter bundles list --keyword mysql
+# Bundles that have the keyword mysql defined
+```
+
+#### Register a Bundle
+
+Bundles are automatically registered in the index when they are pulled.
+A bundle can be manually indexed with the `porter bundle add` command.
+If a bundle was already set as the default, it is unset and the specified bundle becomes the default.
+
+```console
+$ porter bundle add --repository REPOSITORY [--interface INTERFACE]
+# Pulls the bundle and indexes it
+
+$ porter bundle add --repository ghcr.io/getporter/helm-mysql --interface io.cnab.interfaces.mysql
+# Registers the helm-mysql bundle as the default implementation for io.cnab.interfaces.mysql
+```
+
+Note: I chose to use the verb "add" instead of register because it allows us to have consistent verbs across resources. Having a register alias may be useful as well.
+
+
+#### Unregister a Bundle
+
+A bundle can be removed from the bundle index with the `porter bundle remove` command.
+
+```console
+$ porter bundle remove --repository REPOSITORY
+```
+
+⛔️ Bundle index management, such as automatically removing unused bundles, scanning bundles for new versions, or removing all bundles, is out-of-scope of this proposal.
+
+
+### Post-Install Dependency Management
+
+At installation, it's clear that Porter should call the install action for any dependencies that do not already exist.
+After that, each bundle may have different requirements for if the dependencies should be upgraded in lock-step, if the bundle is a decomposed set of services or has looser relationships.
+Currently when a Porter action is executed on a bundle, that action is automatically executed on its dependencies as well.
+In many of the use cases identified in this proposal, that is not the desired behavior.
+
+New configuration at the bundle level and at the point when a bundle action is run should allow people to control this behavior.
+
+All bundle actions (install, upgrade, invoke, uninstall) have a common flag, `--dependencies=true|false`, that let the user override the dependency behavior set in the bundle. This flag controlls whether or not Porter's dependency management step should be executed or not, and defaults to true.
+
+The user can decide to install a bundle without its dependencies, manually passing in data that would have come from the dependency as credentials.
+
+```console
+$ porter install myapp --reference myorg/mybundle --dependencies=false --cred logger
+# the logger credential contains the logger endpoint normally output by the logger dependency
+```
```

Comment:
How to push more parameters from the top bundle to the inner one?

Reply:
This is a known gap and is being tracked by this issue https://github.com/getporter/porter/issues/990.

Reply:
I have updated my answer based on recent changes to the revised CNAB Dependencies spec. See "Wiring Dependencies" section for more details.